daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/CompiledExpression.scala

```diff
@@ -235,6 +236,16 @@ class ExpressionCompiler[T <: AnyRef] extends ExpressionCompilerBase[T] {
       compileInfoWhereExpressionWasLocated.SDE(msg, exprOrLiteral)
     }
 
+    if (expr.contains(".[") && expr.matches(".*\\.\\[.*].*")) {
```

Comment:
I try to avoid string comparisons when possible. Especially with regular expression where it's easy to have unintentional matches. The DFDLExpressionParser already parses steps and predicates and turns them into objects, so it might be better to do this checks in the objects it creates.

For example the, `SelfStepExpression` class in Expression.scala has an `Option[PredicateExpression]`. Can we just do something like:
```scala
if (predArg.isDefined) SDW(...)
```
Do we want the same warning for `UpStepExpression`?